File: backend/Graph.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace TestingBackend {
    public interface IHasId<T>
    {
        public T Id { get; }
    }
    
    public class Graph<N,E> where N : IHasId<Guid>
    {
        private Dictionary<Guid, N> Nodes = new();
        private Dictionary<(Guid, Guid), E> Edges = new();
        public Graph() {}

        public void AddNode(N node)
        {
            Nodes.Add(node.Id, node);
        }

        public void AddEdge(N a, N b, E weight)
        {
            Edges.Add((a.Id, b.Id), weight);
        }

        public IEnumerable<(N Node, E Weight)> GetNeighbours(N node)
        {
            foreach (var ((edge1, edge2), weight) in Edges)
            {
                if (edge1 == node.Id)
                    yield return (Nodes[edge2], weight);

                if (edge2 == node.Id)
                    yield return (Nodes[edge1], weight);
            }
        }
    }
}

File: frontend/src/logic/graph.ts
import { NTuple } from './models/ntuple';
import { Ord, fromCompare } from 'fp-ts/Ord';
import * as N from 'fp-ts/number';

export type Edge = [NTuple, NTuple, number];

export namespace Edge {
  export const Ord: Ord<Edge> = fromCompare((first, second) =>
    N.Ord.compare(first[2], second[2]),
  );
}

export class Graph {
  nodes: NTuple[];
  edges: Edge[];

  constructor(nodes: NTuple[] = [], edges: Edge[] = []) {
    this.nodes = nodes;
    this.edges = edges;
  }

  addNode(node: NTuple) {
    // If it is already in the graph, don't add it
    if (this.nodes.some((x) => NTuple.Eq.equals(x, node))) return;

    this.nodes.push(node);
  }

  addNodes(nodes: NTuple[]) {
    nodes.forEach((x) => this.addNode(x));
  }

  hasEdge(from: NTuple, to: NTuple): boolean {
    return this.edges.some((edge) => this.isEdge(from, to, edge));
  }

  private isEdge(from: NTuple, to: NTuple, edge: Edge): boolean {
    const [x, y, _] = edge;
    const containsThisWay =
      NTuple.Eq.equals(x, from) && NTuple.Eq.equals(y, to);
    const containsThatWay =
      NTuple.Eq.equals(y, from) && NTuple.Eq.equals(x, to);

    return containsThisWay || containsThatWay;
  }

  addEdge(from: NTuple, to: NTuple, weight: number) {
    if (this.hasEdge(from, to)) return;

    this.edges.push([from, to, weight]);
  }

  shallowClone(): Graph {
    const edgesClone = this.edges.map(([x, y, z]) => [x, y, z] as Edge);
    return new Graph(this.nodes, edgesClone);
  }

  getNeighbours(node: NTuple): NTuple[] {
    const neighbours = this.edges
      .map(([from, to, _]) => {
        if (NTuple.Eq.equals(from, node)) return to;

        if (NTuple.Eq.equals(to, node)) return from;

        return null;
      })
      .filter((x) => x !== null);

    return neighbours as NTuple[];
  }

  removeNode(node: NTuple) {
    this.nodes = this.nodes.filter((x) => !NTuple.Eq.equals(x, node));
    this.edges = this.edges.filter(
      ([x, y, _]) => !NTuple.Eq.equals(x, node) && !NTuple.Eq.equals(y, node),
    );
  }
}
